packages/muelu/src/Transfers/Matrix-Free/MueLu_MatrixFreeTentativeP_kokkos_def.hpp

```diff
@@ -0,0 +1,144 @@
+
+    // Step 1: Y*=beta*Y, setup necessary structures
+    Y.scale(beta);
+
+    auto kokkos_view_X = X.getDeviceLocalView(Xpetra::Access::ReadOnly);
+    auto kokkos_view_Y = Y.getDeviceLocalView(Xpetra::Access::ReadWrite);
+    std::cout << "kokkos_view_X.extent(0) = " << kokkos_view_X.extent(0) << std::endl;
+    std::cout << "kokkos_view_Y.extent(0) = " << kokkos_view_Y.extent(0) << std::endl;
+
+    LO numCols = kokkos_view_X.extent(1);
+
+    if(mode == Teuchos::TRANS) { // if we're in restrictor mode
+      LO numAggs = aggregates_->GetNumAggregates();
+      Kokkos::View<LO*,DeviceType> aggPtr;
+      Kokkos::View<LO*,DeviceType> aggNodes;
+      Kokkos::View<LO*,DeviceType> unaggregated;
+      typename Aggregates_kokkos::aggregates_sizes_type::const_type aggSizes = aggregates_->ComputeAggregateSizes(); // TODO: can probably remove this and just use aggPtr if needed
+      aggregates_->ComputeNodesInAggregate(aggPtr,aggNodes,unaggregated);
+
+      // Step 2: Compute Y=Y+alpha*R*X
+      // recall R*X is an average of X over aggregates
+      Kokkos::parallel_for("MueLu:MatrixFreeTentativeR_kokkos:apply", range_type(0,numAggs), 
+        KOKKOS_LAMBDA(const int aggIdx) {
+          LO aggSize = aggSizes(aggIdx);
+          // TODO: consider team policies or alternatives for agg average?
+          // TODO: how about team policies for columns of a multivector?
+          // TODO: probably smarter to sqrt the whole aggSizes once
+          for(int colIdx=0; colIdx<numCols; ++colIdx) {
+            Scalar aggAvg = 0;
+            for(LO i=0; i<aggSize; ++i) {
+              aggAvg += kokkos_view_X(aggNodes(aggPtr(aggIdx)+i),colIdx);
+            }
+            aggAvg /= std::sqrt(static_cast<Scalar>(aggSize)); // scale row of R to have norm 1
+            kokkos_view_Y(aggIdx,colIdx) += alpha*aggAvg;
+          }
+        });
+    } else { // if we're in prolongator mode
+      const auto vertex2Agg = aggregates_->GetVertex2AggId();
+      auto vertex2AggView = vertex2Agg->getDeviceLocalView(Xpetra::Access::ReadOnly);
+      typename Aggregates_kokkos::aggregates_sizes_type::const_type aggSizes = aggregates_->ComputeAggregateSizes();
+
+      // Step 2: Compute Y=Y+alpha*P*X
+      // recall P*X is essentially injection of X, but sum if a node belongs to multiple aggregates
+      Kokkos::parallel_for("MueLu:MatrixFreeTentativeP_kokkos:apply", range_type(0,kokkos_view_Y.extent(0)), 
+        KOKKOS_LAMBDA(const int fineIdx) {
+          LO aggIdx = vertex2AggView(fineIdx,0);
+          
+          // TODO: how about team policies for columns of a multivector?
+          // TODO: probably smarter to sqrt the whole aggSizes once
+          for(int colIdx=0; colIdx<numCols; ++colIdx)
+            kokkos_view_Y(fineIdx,colIdx) += alpha*kokkos_view_X(aggIdx,colIdx)/sqrt(aggSizes(aggIdx));
+        });
+    }
+  } 
+
+  // I don't care
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class DeviceType>
```

Comment:
This should probably just throw cause it doesn't make much sense.

Reply:
Yeah... it's an pure virtual function, so I have to override it, and it's a one-liner to do correctly, but I don't know if we use that interface anywhere in MueLu. This was my solution that does something correct despite not being needed.

Reply:
What's the residual supposed to be for a square matrix? I'd just throw.

Reply:
oh oh good point. I strictly put the equation there without thinking about it. done!